### ISSUE TITLE
Format string injection vulnerability

### ISSUE THREAD

**Issue by Arinerron** [sickill/stderred#80]
Hey, I have some really weird format string injection-type behavior going on that I think I can attribute to libstderrred.

![image](https://user-images.githubusercontent.com/3926753/100692037-d6394b00-333e-11eb-8c3e-fa6febc244b1.png)

It's definitely not just the `ls` binary though:

![image](https://user-images.githubusercontent.com/3926753/100692284-5c559180-333f-11eb-9c45-090f683207ba.png)

Potentially useful info:
```
$ command -v ls
alias ls='ls --color=tty'

$ env | grep LD_PRELOAD
LD_PRELOAD=/usr/lib/libstderred.so:/usr/lib/libstderred.so

$ trizen -Q stderred 
stderred-git v1.0.r37.g399e3b1-1

$ LD_PRELOAD= ls '%x'
ls: cannot access '%x': No such file or directory

$ strace ls '%x %x %x' 2>&1 | grep 'write'
write(2, "ls: ", 4ls: )                     = 4
write(2, "cannot access '4710e010 47114ab0"..., 42cannot access '4710e010 47114ab0 6bf27a60') = 42
write(2, ": No such file or directory", 27: No such file or directory) = 27
write(2, "\n", 1
```

Let me know if you can think of any other relevant information you want. Thanks!

**Comment by Arinerron** [comment]
According to the backtrace, libstderrred's `new_do_write` (at least) is involved

![image](https://user-images.githubusercontent.com/3926753/100692623-15b46700-3340-11eb-86c5-0309eb2f13d7.png)

**Comment by Arinerron** [comment]
It looks like it's the wrapper functions for
* [`error`](https://github.com/sickill/stderred/blob/master/src/stderred.c#L244)
* [`error_at_line`](https://github.com/sickill/stderred/blob/master/src/stderred.c#L269)
that are evaluating format strings twice.

**Comment by DidierA** [comment]
Hi,
I have the same issue, and can confirm it only happens when stderred is used.
A small C program to reproduce this bug:
```
#include <error.h>

int main(int ac, char *av[]) {
    error(0,22, "test %s", "%x") ;
    return 0;
}
```
output with LD_PRELOAD not set:  
`./test_error: test %x: Invalid argument`

output with LD_PRELOAD set to use stderred:  
`./test_error: test 41c7e2f0: Invalid argument`

**Comment by DidierA** [comment]
Fixed in my fork. Please check https://github.com/DidierA/stderred/commit/d105de8de47e3792bd1b1d50e3375e59260c5b96

**Comment by cehoffman** [comment]
That looks good to me. If you’re willing to submit a PR Is be happy to accept it.

**Comment by DidierA** [comment]
Done, PR #81

**Comment by Arinerron** [comment]
I can confirm it's fixed as of 2c41540:

![image](https://user-images.githubusercontent.com/3926753/104071071-37b1cc80-51bd-11eb-94e2-a0b73e3f2029.png)
